tests/plugin/test_http_proxy_plugins.py

```diff
@@ -56,6 +59,14 @@ def setUp(self,
             flags=self.flags)
         self.protocol_handler.initialize()
 
+    def tearDown(self) -> None:
+        tmpDir = Path(tempfile.gettempdir())
+        for f in tmpDir.glob('proxy-cache-*'):
+            if f.is_file():
+                os.remove(f)
+        if tmpDir.joinpath('list.txt').is_file():
```

Comment:
During parallel testing, `os.remove` results in a deadlock as `list.txt` is also being used by other tests. See https://github.com/abhinavsingh/proxy.py/pull/425/checks?check_run_id=1197597223

Reply:
Ah OK! I did not know CI was using parallel testing. When I run them locally, I think they execute sequentially. Am I right?
I tried to investigate this issue but I could not understand what was happening (I am under Linux and have no ways of running Windows test locally). 

Actually, multi-threading is already a major issue with the disk-based cache, since I have to find a way to synchronize file accesses. I used a `multiprocessing.Lock()`, but I am not sure it is the best solution out there. Indeed, the reader can access the file simultaneously, while the writes must access the file sequentially. A "read-write lock" would be more appropriate. Or we could go for OS file locks, but that would be platform-dependent.
For multi-processing, I am not sure what we can do, since `multiprocessing.Lock()` will be insufficient, as it is not aware of the other instances of tests running in parallel, right?

Reply:
> Ah OK! I did not know CI was using parallel testing. When I run them locally, I think they execute sequentially. Am I right?

Yes, tests run sequentially.  Thinking about it further, as same command invocation is used for CI, there is no reason they should run in parallel on GitHub actions.

> For multi-processing, I am not sure what we can do, since multiprocessing.Lock() will be insufficient, as it is not aware of the other instances of tests running in parallel, right?

Yes.  And idea of shared information about other tests doesn't sound right in the first place.  There are some sporadic issues related to `multiprocessing` and `windows` too.  I have a windows vm locally, will give it a try.

We can bypass tests conditionally. Example

```
@unittest.skipIf(
    os.name == 'nt',
    'Open file limit tests disabled for Windows')
class TestSetOpenFileLimit(unittest.TestCase):
```

> Actually, multi-threading is already a major issue with the disk-based cache, since I have to find a way to synchronize file accesses.

Yes.  A single file will require synchronization.  Avoiding lock/synchronizations can also boost the cache storage performance by many folds.  `list.txt` is essentially an index file.  We can even rename it appropriately in later iterations.  It contains metadata about stored item in a random order i.e. stored indexes are not ordered by any attribute.  We can also see `list.txt` as serialized index store as indexes exists in the order of requests.

Currently, we don't have an update flow i.e. once written, cache is immutable and never updated.  But, design wise we must prepare for an update cycle :) where cache is updated either based upon time based expiry, query parameters etc.  This decision making can be moulded into cache plugin lifecycle callback.

> I used a `multiprocessing.Lock()`, but I am not sure it is the best solution out there. Indeed, the reader can access the file simultaneously, while the writes must access the file sequentially. A "read-write lock" would be more appropriate. Or we could go for OS file locks, but that would be platform-dependent.

The task of on disk based storage leads us into database designing wonderland :) Good thing is our cache storage is pluggable and we can easily add new storage mechanisms _(leveldb, sqlite, cloud storages, ....)_

Few thoughts around approaching disk based storage using zero-dependency philosophy:

1) Storage:  We are currently caching responses as a single file per request.  So storage becomes pretty straightforward.  But for a long running server or with lots of data, I do expect us to run into situations where OS starts to run out of file descriptors _(possibly fixable by tweaking kernel parameters)_.  Let's not bother about it for now as fixing this is relatively easier if we get our indexes right.

2) Indexes:  Without any ordering on stored indexes, we indirectly increase one or both:
a) Startup time:  Cache plugin is busy ordering the indexes in-memory (if we decide to pre-warm the cache index).
b) Lookup time:  Cache plugin is busy iterating over unordered indexes (if we decide to consult index file for every read).

In short, ordering (fully or partially) indexes on disk can greatly benefit overall performance.  Thinking of primitive ways :) maintaining one index file per alphabet e.g. `a-list.txt`, `b-list.txt`, .... where each file store data about domains starting with respective alphabet (or any unicode character).  Such partitioning strategies can greatly reduce the lock contention across workers.

At some point we can also use an in-memory cache e.g. LRU cache to minimize disk hits.  Ideally, I would like to have a running copy of indexes in-memory but that is not always practical with lots of indexes (and not enough available RAM).

3) Synchronization: There is in-build `pubsub` mechanism in `proxy.py`.  Try it independently here https://github.com/abhinavsingh/proxy.py/blob/develop/examples/pubsub_eventing.py.

For some background, eventing mechanism was developed for `devtools` integration.  This allows `proxy.py` traffic to be inspected using `chrome devtools`.  `devtools` expects a `websocket` connection sending stream of events as described in `devtools` protocol specification.  In short, we do have a mechanism to produce a single stream of custom events from various worker processes.  Most importantly, eventing mechanism relieves us from synchronization pains.  

An idea could be to simply publish cache payload as events from within cache plugin.  Then, a single global subscriber thread or multiple subscriber threads (one per worker process) can consume events and write indexes to disk either synchronously (as events are received) or in batches.

Hopefully, we can mix some of the above ideas to build our own storage layer :)

Reply:
> We can bypass tests conditionally. Example
> 
> ```
> @unittest.skipIf(
>     os.name == 'nt',
>     'Open file limit tests disabled for Windows')
> class TestSetOpenFileLimit(unittest.TestCase):
> ```

Yes, I knew we can do this, but I wanted you to see if you better understands what happens on Windows. The current output from the tests is a bit short and I did not want to push tests commits only to see a more verbose output.

> Yes. A single file will require synchronization. Avoiding lock/synchronizations can also boost the cache storage performance by many folds. `list.txt` is essentially an index file. We can even rename it appropriately in later iterations. It contains metadata about stored item in a random order i.e. stored indexes are not ordered by any attribute. We can also see `list.txt` as serialized index store as indexes exists in the order of requests.

Exactly, that is how I designed it.
 
> Currently, we don't have an update flow i.e. once written, cache is immutable and never updated. But, design wise we must prepare for an update cycle :) where cache is updated either based upon time based expiry, query parameters etc. This decision making can be moulded into cache plugin lifecycle callback.

Yes, I kept that for later, because I think that there is quite a large amount of work to implement all cache expiracy features.

> The task of on disk based storage leads us into database designing wonderland :) Good thing is our cache storage is pluggable and we can easily add new storage mechanisms _(leveldb, sqlite, cloud storages, ....)_
> 
> Few thoughts around approaching disk based storage using zero-dependency philosophy:
> 
> 1. Storage:  We are currently caching responses as a single file per request.  So storage becomes pretty straightforward.  But for a long running server or with lots of data, I do expect us to run into situations where OS starts to run out of file descriptors _(possibly fixable by tweaking kernel parameters)_.  Let's not bother about it for now as fixing this is relatively easier if we get our indexes right.

I do not think we will run out of file descriptors because there is only one open file per non-completed request plus the index. This should not be too many for the OS, unless we open a lot of requests in parallel on purpose.
Though, if we use it for a long time we will run into the limit of inodes per filesystem. For Ext2/3/4 filesystems, it can be configured when formatting the disk. To avoid this we can use subfolders as is done in e.g. `.git/objects` "cache". 

> 2. Indexes:  Without any ordering on stored indexes, we indirectly increase one or both:
>        a) Startup time:  Cache plugin is busy ordering the indexes in-memory (if we decide to pre-warm the cache index).
>        b) Lookup time:  Cache plugin is busy iterating over unordered indexes (if we decide to consult index file for every read).
> 
> 
> In short, ordering (fully or partially) indexes on disk can greatly benefit overall performance. Thinking of primitive ways :) maintaining one index file per alphabet e.g. `a-list.txt`, `b-list.txt`, .... where each file store data about domains starting with respective alphabet (or any unicode character). Such partitioning strategies can greatly reduce the lock contention across workers.

Yes, this will become an issue for large caches. The list by alphabet letter seems a good compromise, but we have to get a reliable way to determine it, without having everything in `w` (many websites still use `www`). Or we could even go for a per-host index file. I do not know how fast is `os.path.is_file()`.